File: Universe_Backend/Data/Models/Post.cs
using System.ComponentModel.DataAnnotations.Schema;
using Universe_Backend.Data.DTOs;

namespace Universe_Backend.Data.Models;

public class Post
{
    public required int Id { get; set; }
    public required string Uid { get; set; }
    public required string? Title { get; set; }
    public required string Body { get; set; }
    public List<string> Images { get; set; } = [];
    public List<string> Videos { get; set; } = [];
    public List<string> Audios { get; set; } = [];
    public DateTime PublishDate { get; set; } = DateTime.Now;
    public int? ReplyToPostId { get; set; }
    public int? ChildPostId { get; set; }
    public required string AuthorId { get; set; }
    public ICollection<Widget>? Widgets { get; set; }
    public int ReactionsCount { get; set; }
    public int RepliesCount { get; set; }

    public virtual User? Author { get; set; }
    public virtual Post? ReplyToPost { get; set; }
    public virtual Post? ChildPost { get; set; }
    public virtual ICollection<PostReaction> Reactions { get; set; } = [];
    public virtual ICollection<Tag> Tags { get; set; } = [];
    public virtual ICollection<User> Mentions { get; set; } = [];

    public PostDTO ToDTO()
    {
        return new PostDTO
        {
            Id = Id,
            Uid = Uid,
            Title = Title,
            Body = Body,
            Images = Images,
            Videos = Videos,
            Audios = Audios,
            PublishDate = PublishDate,
            ReplyToPostId = ReplyToPostId,
            ChildPostId = ChildPostId,
            Widgets = Widgets,
            Author = Author!.ToDTO(),
            ReactionsCount = ReactionsCount,
            RepliesCount = RepliesCount,
            ReplyToPost = ReplyToPost?.ToDTO(),
            ChildPost = ChildPost?.ToDTO()
        };
    }

    public Post UpdateFromDTO(PostDTO post)
    {
        Title = post.Title;
        Body = post.Body;
        Images = post.Images;
        Videos = post.Videos;
        Audios = post.Audios;
        PublishDate = post.PublishDate;
        ReplyToPostId = post.ReplyToPostId;
        ChildPostId = post.ChildPostId;
        AuthorId = post.Author.Id;
        Widgets = post.Widgets;
        ReactionsCount = post.ReactionsCount;
        RepliesCount = post.RepliesCount;
        return this;
    }
}

File: universe/lib/models/data/post.dart
import 'package:universe/models/data/post_reaction.dart';
import 'package:universe/models/data/user.dart';
import 'package:universe/models/data/widget.dart';
import 'package:json_annotation/json_annotation.dart';
import 'package:uuid/uuid.dart';

part 'post.g.dart';

@JsonSerializable()
class Post {
  int id;
  String? uid;
  final String title;
  final String body;
  List<dynamic> images;
  List<dynamic> videos;
  List<dynamic> audios;
  final DateTime publishDate;
  final int replyToPostId;
  final int childPostId;
  final List<Widget> widgets;
  final User author;
  int reactionsCount;
  int repliesCount;
  bool reactedToByCaller;
  PostReaction? callerReaction;
  Post? childPost;
  Post? replyToPost;

  Post({
    this.id = -1,
    this.uid,
    required this.title,
    required this.body,
    required this.images,
    required this.videos,
    required this.audios,
    required this.publishDate,
    this.replyToPostId = -1,
    this.childPostId = -1,
    required this.widgets,
    required this.author,
    required this.reactionsCount,
    required this.repliesCount,
    required this.reactedToByCaller,
    this.childPost,
    this.replyToPost,
  }) {
    uid ??= const Uuid().v4();
  }

  Post copyWith({
    int? id,
    String? uid,
    String? title,
    String? body,
    List<dynamic>? images,
    List<dynamic>? videos,
    List<dynamic>? audios,
    DateTime? publishDate,
    int? replyToPostId,
    int? childPostId,
    List<Widget>? widgets,
    User? author,
    int? reactionsCount,
    int? repliesCount,
    bool? reactedToByCaller,
    PostReaction? callerReaction,
  }) {
    return Post(
      id: id ?? this.id,
      uid: uid ?? this.uid,
      title: title ?? this.title,
      body: body ?? this.body,
      images: images ?? this.images,
      videos: videos ?? this.videos,
      audios: audios ?? this.audios,
      publishDate: publishDate ?? this.publishDate,
      replyToPostId: replyToPostId ?? this.replyToPostId,
      childPostId: childPostId ?? this.childPostId,
      widgets: widgets ?? this.widgets,
      author: author ?? this.author,
      reactionsCount: reactionsCount ?? this.reactionsCount,
      repliesCount: repliesCount ?? this.repliesCount,
      reactedToByCaller: reactedToByCaller ?? this.reactedToByCaller,
    )..callerReaction = callerReaction ?? this.callerReaction;
  }

  factory Post.fromJson(Map<String, dynamic> json) => _$PostFromJson(json);

  Map<String, dynamic> toJson() => _$PostToJson(this);
}
